pyfar/dsp/dsp.py

```diff
@@ -2669,3 +2669,171 @@ def normalize(signal, reference_method='max', domain='auto',
+        matches the cshape to which `signal_1` and `signal_2` were broadcasted.
+        The time lags are given in seconds and can be converted to samples by
+        multiplication with ``signal_1.sampling_rate``. In this case, they are
+        in the interval ``[-signal_2.n_samples + 1, signal_1.n_samples - 1]``
+        if the full correlation was computed. In case of the cyclic
+        correlation, they are in the interval
+        ``[-(signal_1.n_samples // 2) + 1, signal_1.n_samples // 2]``
+        if the signals have an even number of samples, and in the interval
+        ``[-(signal_1.n_samples // 2), signal_1.n_samples // 2]``
+        if the signals have an odd number of samples.
+
+    Examples
+    --------
+
+    Linear and cyclic auto-correlation of a perfect sequence. Perfect sequences
+    have unit auto-correlation for :math:`l=0` and zero auto-correlation
+    otherwise
+
+    .. plot::
+
+        >>> import pyfar as pf
+        >>>
+        >>> signal = pf.signals.linear_perfect_sweep(2**8)
+        >>>
+        >>> for mode in ['full', 'cyclic']:
+        >>>     cor = pf.dsp.correlate(signal, signal, mode)
+        >>>     ax = pf.plot.time(cor / pf.dsp.energy(signal), label=mode)
+        >>>
+        >>> ax.set_xlabel('time lags in seconds')
+        >>> ax.set_ylabel('auto correlation')
+        >>> ax.legend()
+
+    Compute the lags (delay) that are required to maximize the
+    cross-correlation between a one and multi-dimensional signal
+
+    .. plot::
+
+        >>> import pyfar as pf
+        >>> import numpy as np
+        >>>
+        >>> # one-dimensional signal: impulse with zero delay
+        >>> signal_1 = pf.signals.impulse(5, 0)
+        >>>
+        >>> # multi-dimension signal with cshape = (2, 2):
+        >>> # impulses with non-zero delays
+        >>> delays = np.array([[0, 1], [2, 3]], dtype=int)
+        >>> signal_2 = pf.signals.impulse(5, delays)
+        >>>
+        >>> cor = pf.dsp.correlate(signal_1, signal_2, 'full')
+        >>>
+        >>> # compute the lags in samples
+        >>> argmax = cor.times[np.argmax(cor.time, axis=-1)]
+        >>> argmax *= signal_1.sampling_rate
+
+    In this case the lags correspond to the negative delays:
+    ``argmax = [[0, -1], [-2, -3]]``.
+    """  # noqa: E501
+
+    # check input
+    if type(signal_1) is not pyfar.Signal or \
+            type(signal_2) is not pyfar.Signal:
+        raise TypeError("signal_1 and signal_2 must be pyfar.Signal objects")
+
+    if signal_1.sampling_rate != signal_2.sampling_rate:
+        raise ValueError("Both signals must have the same sampling rate.")
+
+    if mode not in ['full', 'cyclic']:
+        raise ValueError(f"mode is '{mode}' but must be 'full' or 'cyclic'")
+
+    # copy input to avoid changing mutual data
+    signal_1 = signal_1.copy()
+    signal_2 = signal_2.copy()
+
+    # check for complex signals
+    if signal_1.complex and not signal_2.complex:
```

Comment:
we should mention this in the doc at some point

Reply:
we could tell it in the output definition, is complex if one of the input singals is complex. many other thigs such as broadcasting is also expected, so besser haben wie brauchen.

Reply:
Good point, added it there.